keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+and standardization by Kubernetes. This could be used to describe topology
+across resources from different vendors, but this is out-of-scope for now.
+
+If a driver needs to reduce resource capacity, then there is a risk that a
+claim gets allocated using that capacity while the kubelet is updating a
+NodeResourceSlice. The implementations of semantic models must handle scenarios
+where more resources are allocated than available. The kubelet plugin of a DRA
+driver ***must*** double-check that the allocated resources are still available
+when NodePrepareResource is called. If not, the pod cannot start until the
+resource comes back. Treating this as a fatal error during pod admission would
+allow us to delete the pod and trying again with a new one.
+
+### Using semantic parameters
+
+The following is an example CRD which the developer of the
+`cards.dra.example.com` DRA driver might define as a valid claim parameters
+object for requesting access to its GPUs:
+
+```yaml
+kind: CardParameters
+apiVersion: dra.example.com/v1alpha1
+metadata:
+  name: my-parameters
+  namespace: user-namespace
+  uid: foobar-uid
+...
+spec:
+  count: 2
+  minimumRuntimeVersion: v12.0.0
+  minimumMemory: 8Gi
+  # "sharing" is a configuration parameter that does not
+  # get translated into the selector below.
+  sharing:
+    strategy: TimeSliced
+```
+
+Note that all fields in this CRD can be fully validated since it is owned by
+the DRA driver itself. This includes value ranges that are specific to the
+underlying hardware. There's no risk of using invalid attribute names because
+only the fields shown here are valid.
+
+With this CRD in place, a DRA driver controller is able to convert instances of
+it into a generic, in-tree `ResourceClaimParameters` object that the scheduler
+is able to understand.
+
+For the example above, the converted object would look as follows:
+
+```yaml
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+
+metadata:
+  # This cannot be the same as my-parameters because parameter objects with a different
+  # type might also use it. Instead, the original object gets linked to below.
+  name: someArbitraryName
+  namespace: user-namespace
+
+generatedFrom:
+  name: my-parameters
+  kind: CardParameters
+  apiGroup: dra.example.com
+  uid: foobar-uid
+
+vendorParameters:
+  # A vendor can put any kind of object here to pass the configuration
+  # parameters down to the kubelet plugin. In this case, the vendor
+  # driver simply copied the entire CR. It could also be some
+  # separate, smaller configuration type.
+  kind: CardParameters
+  apiVersion: dra.example.com/v1alpha1
+  metadata:
+    name: my-parameters
+    namespace: user-namespace
+    uid: foobar-uid
+  ...
+  spec:
+    count: 2
+    minimumRuntimeVersion: v12.0.0
+    minimumMemory: 8Gi
+    sharing:
+      strategy: TimeSliced
+
+requests:
+- driverName: cards.dra.example.com
+  partitionable:
```

Comment:
So, the API is actually expressive enough such that a single driver could publish NodeResourceCapacity with different models, as long as they were in different NodeResourceCapacity objects. Alternatively we could even relax the "union" and make it so the pointers are not mutually exclusive. Is that something we want? If we do that, then in the requests, can they span multiple models (but not drivers still).

I don't know that we need to answer that in 1.30, but it's something to think about.

Reply:
I agree. I don't see the need to artificially restrict a driver to only advertising resources with a single model (so long as each model governs a _distinct_ set of resources).

Reply:
> So, the API is actually expressive enough such that a single driver could publish NodeResourceCapacity with different models, as long as they were in different NodeResourceCapacity objects. 

s/NodeResourceCapacity/NodeResourceSlice/ ... :smile: 

Publishing different objects with one model per object was what I was thinking when writing down the Go type. It keeps the type simpler.

At some point, "As mentioned previously, exactly one semantic model must be used by a given driver." was added. That's actually not a hard rule and not enforced by the API. I'll double-check where it all occurs and replace with the opposite: "drivers can use different models as long as each model represents a  distinct set of resources".

> Alternatively we could even relax the "union" and make it so the pointers are not mutually exclusive.

I like that less because "one-of-many" is easier to reason about when it's done consistently.